cirq/experiments/qubit_characterizations.py

```diff
@@ -0,0 +1,544 @@
+    sigma_z are the Identity and Pauli matrices.
+
+    Based on the measured probabilities probs and the transformations of the
+    measurement operator by different basis rotations, one can build an
+    overdetermined set of linear equations numpy.dot(mat, c) = probs. Here c
+    is of length 15 and contains all the c_ij's (except c_00 which is set to
+    1), and mat is a 27 by 15 matrix having three non-zero elements in each
+    row that are either 1 or -1.
+
+    The least-square solution to the above set of linear equations is used to
+    construct the density matrix rho.
+
+    See Vandersypen and Chuang, Rev. Mod. Phys. 76, 1037 for details.
+
+    Args:
+        sampler: The quantum engine or simulator to run the circuits.
+        q_0: The first qubit under test.
+        q_1: The second qubit under test.
+        circuit: The circuit to execute on the qubits before tomography.
+        num_shots: The number of measurements for each basis rotation.
+
+    Returns:
+        A StateTomographyResult object that stores and plots the density matrix.
+    """
+
+    def _measurement(two_qubit_circuit: circuits.Circuit) -> Sequence[float]:
+        two_qubit_circuit.append(ops.measure(q_0, q_1, key='z'))
+        results = sampler.run(two_qubit_circuit, repetitions=num_shots)
+        bit_strings = [r for r in results.measurements['z']]
+        p_00 = 0.0
+        p_01 = 0.0
+        p_10 = 0.0
+        for bits in bit_strings:
+            if not bits[0] and not bits[1]:
+                p_00 += 1.0 / num_shots
+            elif not bits[0] and bits[1]:
+                p_01 += 1.0 / num_shots
+            elif bits[0] and not bits[1]:
+                p_10 += 1.0 / num_shots
+        return [p_00, p_01, p_10]
+
+    sigma_0 = numpy.eye(2) / 2.0
+    sigma_1 = numpy.array([[0.0, 1.0], [1.0, 0.0]]) / 2.0
+    sigma_2 = numpy.array([[0.0, -1.0j], [1.0j, 0.0]]) / 2.0
+    sigma_3 = numpy.array([[1.0, 0.0], [0.0, -1.0]]) / 2.0
+    sigmas = [sigma_0, sigma_1, sigma_2, sigma_3]
+
+    probs = []
+    rots = [ops.X ** 0, ops.X ** 0.5, ops.Y ** 0.5]
+    mat = numpy.zeros((27, 15))
+    s = numpy.array([[1.0, 1.0, 1.0], [-1.0, 1.0, -1.0], [1.0, -1.0, -1.0]])
+
+    for i, rot_1 in enumerate(rots):
+        for j, rot_2 in enumerate(rots):
+            m_idx, indices, signs = _indices_after_basis_rot(i, j)
+            mat[m_idx: (m_idx + 3), indices] = s * numpy.tile(signs, (3, 1))
+            test_circuit = circuit + circuits.Circuit.from_ops(rot_1(q_1))
+            test_circuit.append(rot_2(q_0))
+            probs.extend(_measurement(test_circuit))
+
+    c, _, _, _ = numpy.linalg.lstsq(mat, 4.0 * numpy.array(probs) - 1.0,
+                                    rcond=None)
+    c = numpy.concatenate(([1.0], c))
+    c = c.reshape(4, 4)
+
+    rho = numpy.zeros((4, 4))
+    for i in range(4):
+        for j in range(4):
+            rho = rho + c[i, j] * numpy.kron(sigmas[i], sigmas[j])
```

Comment:
`+=`

Reply:
Got error: Cannot cast ufunc add output from dtype('complex128') to dtype('float64') with casting rule 'same_kind'.

I can in principle do `rho += (c[i, j] * numpy.kron(sigmas[i], sigmas[j])).astype('complex')` but it seems more complicated